core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java

```diff
@@ -355,11 +355,85 @@ public static boolean regexpContains(String value, String regex) {
       Pattern regexp = Pattern.compile(regex);
       return regexp.matcher(value).find();
     } catch (PatternSyntaxException ex) {
-      throw RESOURCE.invalidInputForRegexpContains(ex.getMessage().replace("\r\n", " ")
-          .replace("\n", " ").replace("\r", " ")).ex();
+      throw RESOURCE.invalidRegexInputForRegexpFunctions(ex.getMessage().replace("\r\n", " ")
+              .replace("\n", " ").replace("\r", " "),
+          "REGEXP_CONTAINS").ex();
     }
   }
 
+  /** SQL {@code REGEXP_EXTRACT(value, regexp[, position[, occurrence]])} function.
+   *  Returns NULL if there is no match, or if position or occurrence are beyond range.
+   *  Returns an exception if regex, position or occurrence are invalid.*/
+  public static @Nullable String regexpExtract(String value, String regex, Integer... params) {
```

Comment:
is extract just a synonym? you can add it as an alias in `StandardConvertletTable` and then you can refactor/remove a lot this code. You shouldn't need a method for both functions here if they're the same. Let me know if you need more info.